### PR TITLE
chore: git mod - adding init saga and connecting components to ctx

### LINKED ISSUE #36814: [Task] Git mod - Create Git Context Provider

### LINKED ISSUE #37800: [Task] Git mod - Components - Layouts & Modals

### PR DESCRIPTION
**Pull request by brayn003**
## Description
- Adds more selectors
- Adds more sagas
- Introduces init steps for git
- Improvements upon CtxAwareGitQuickActions

Fixes #37800 
Fixes #36814 

## Automation

/ok-to-test tags="@tag.Git"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12295168481>
> Commit: ea1ab497cad677aac36f044462e623e526d421d1
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12295168481&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Git`
> Spec:
> <hr>Thu, 12 Dec 2024 12:06:09 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
  - Introduced a context provider for managing Git-related state.
  - Added a `GitQuickActions` component for various Git actions.
  - Implemented custom hooks for managing Git branches, metadata, operations, and settings.
  - Added Redux Saga functions for fetching Git metadata and protected branches.
  - Enhanced autocommit functionality with polling for progress.
  - Introduced actions for managing the autocommit process and fetching protected branches.
  - Added new actions to initialize Git for editor context and manage loading states.

- **Bug Fixes**
  - Improved error handling in various action creators and sagas.

- **Chores**
  - Updated action creators and types for better type safety and clarity.
  - Refined test cases for components to ensure accurate button selection.
  - Modified request functions to utilize Axios promises directly for better consistency.

### CODE REVIEW

File: app/client/src/git/requests/fetchBranchesRequest.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ import type { AxiosPromise } from "axios";
 
 export default async function fetchBranchesRequest(
   branchedApplicationId: string,
-  params: FetchBranchesRequestParams,
+  params: FetchBranchesRequestParams = { pruneBranches: true },
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/requests/fetchBranchesRequest.ts:11]: doesn't make sense

File: app/client/src/git/components/GitContextProvider/hooks/useGitBranches.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+export default function useGitBranches({
+  artifactType,
+  baseArtifactId,
+}: UseGitBranchesParams): UseGitBranchesReturnValue {
+  const basePayload = useMemo(
+    () => ({ artifactType, baseArtifactId }),
+    [artifactType, baseArtifactId],
+  );
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/components/GitContextProvider/hooks/useGitBranches.ts:43]: its already done!

File: app/client/src/git/components/GitQuickActions/helpers/getPullButtonStatus.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  isPullFailing: boolean;
+  statusBehindCount: number;
+}
+
+const getPullBtnStatus = ({
+  isProtectedMode = false,
+  isPullFailing = false,
+  isStatusClean = false,
+  statusBehindCount = 0,
+}: GetPullButtonStatusParams) => {
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/components/GitQuickActions/helpers/getPullButtonStatus.ts:21]: nope

File: app/client/src/git/components/GitQuickActions/BranchButton/index.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+          source: "BOTTOM_BAR_ACTIVE_BRANCH_NAME",
+        });
+      }
+    },
+    [setIsOpen],
+  );
+
+  useEffect(function ellipsisCheck() {
+    setIsEllipsis(isEllipsisActive(labelTarget.current) ?? false);
+  }, []);
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/components/GitQuickActions/BranchButton/index.tsx:74]: nope

File: app/client/src/git/components/GitQuickActions/QuickActionButton.tsx [hunk 5]
```diff
@@ -73,6 +68,7 @@ function QuickActionButton({
               isDisabled={disabled}
               isIconButton
               kind="tertiary"
+              onClick={onClick}
```
Comment (ashit-rath) [inline app/client/src/git/components/GitQuickActions/QuickActionButton.tsx:71]: Can we give a `data-testid` and use? I think with these changes some cypress modifications would be required as well

File: app/client/src/git/sagas/fetchGitMetadataSaga.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import type { GitArtifactPayloadAction } from "git/store/types";
+import { call, put } from "redux-saga/effects";
+import { validateResponse } from "sagas/ErrorSagas";
+
+export default function* fetchGitMetadataSaga(
+  action: GitArtifactPayloadAction,
+) {
+  const { artifactType, baseArtifactId } = action.payload;
+  const basePayload = { artifactType, baseArtifactId };
+  let response: FetchGitMetadataResponse | undefined;
```
Comment (ashit-rath) [inline app/client/src/git/sagas/fetchGitMetadataSaga.ts:13]: Why define the variable here with `let` and not use `const response` in the yield call?

File: app/client/src/git/sagas/triggerAutocommitSaga.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+function isAutocommitHappening(
+  responseData:
+    | TriggerAutocommitResponseData
+    | FetchAutocommitProgressResponseData
+    | undefined,
+): boolean {
+  return (
+    !!responseData &&
+    (responseData.autoCommitResponse === AutocommitStatus.PUBLISHED ||
```
Comment (ashit-rath) [inline app/client/src/git/sagas/triggerAutocommitSaga.ts:43]: Suggestion: Can we do something like?
```
// define
WHITELISTED = [AutocommitStatus.PUBLISHED, AutocommitStatus.IN_PROGRESS, ...]

// usage
WHITELISTED.includes(responseData.autoCommitResponse)
```
It's cleaner and easier to maintain

File: app/client/src/git/sagas/triggerAutocommitSaga.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+        if (isValidResponse && !isAutocommitHappening(progressResponse?.data)) {
+          yield put(gitArtifactActions.pollAutocommitProgressStop(basePayload));
+        }
+
+        if (!isValidResponse) {
+          yield put(gitArtifactActions.pollAutocommitProgressStop(basePayload));
+        }
+
+        yield delay(1000);
```
Comment (ashit-rath) [inline app/client/src/git/sagas/triggerAutocommitSaga.ts:91]: Can we put the 1000 in a const to define it's meaning?

File: app/client/src/git/components/GitQuickActions/QuickActionButton.tsx (hunk 5 above)
Comment (brayn003) [reply app/client/src/git/components/GitQuickActions/QuickActionButton.tsx:71]: I agree with you. However, let's pick it up as an independent task outside of modularisation. Don't want to disturb the CyTests right now as we are working on a tight timeline

File: app/client/src/git/sagas/fetchGitMetadataSaga.ts (hunk 6 above)
Comment (brayn003) [reply app/client/src/git/sagas/fetchGitMetadataSaga.ts:13]: I am intending to use the response for error handling in the catch block. It's not there in the PR but I will be visiting error handling in future PRs

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**